CRM/Core/BAO/SchemaHandler.php

```diff
@@ -135,9 +135,10 @@ public static function buildFieldSQL($params, $separator, $prefix = ''): string
     $sql .= $prefix;
     $sql .= "`{$params['name']}` {$params['type']}";
 
-    if (!empty($params['required'])) {
-      $sql .= " NOT NULL";
-    }
+    // explicitly set NULL attribute for non-required fields to work around
+    // MySQL's special handling of timestamp columns
+    // see https://dev.mysql.com/doc/refman/8.4/en/timestamp-initialization.html
+    $sql .= empty($params['required']) ? ' NULL' : ' NOT NULL';
```

Comment:
I think this is right, just noting we have two other competing variations:

* This one which is slightly different: https://github.com/civicrm/civicrm-core/blob/024c8b7e61e439b8d200f057ddb6f95fc4762fd7/xml/templates/schema.tpl#L35.
* This one which is yet another slightly different logic: https://github.com/civicrm/civicrm-core/blob/024c8b7e61e439b8d200f057ddb6f95fc4762fd7/mixin/lib/civimix-schema/src/SqlGenerator.php#L164-L170

(The second one is related to https://github.com/civicrm/civicrm-core/pull/31104, which is a different issue but is also about the "first timestamp" column.)

I'm just thinking we should pick _one_ so that the same schema definition doesn't generate different DDL in different situations.

Reply:
@demeritcowboy so the difference between the two is that variant 1 adds the `null` keyword to every non-required field and variant 2 only does so for date columns?
And how is variant 1 different from what's in this PR?

I don't know which is more correct, but it seems harmless to always add it regardless of the field type?

Reply:
Yeah I think I like the version in this PR the best - I just don't know offhand if there was some reason it was qualified in the SqlGenerator version. Both mysql and mariadb docs suggest it's always valid DDL to put null or not null.

The difference with the first variation is that that one doesn't output anything if `{$field.required}` is (boolean-ish) false, or missing. So it would never output NULL unless it's explicitly set to string "false", and it would never output NOT NULL unless explicitly set to string "true", or "kwyjibo", or 5.

Reply:
Ok, sounds good. Here's a PR for the 2nd. https://github.com/civicrm/civicrm-core/pull/31206

Looks like we don't need to worry about the first, as it's only called from this class:
https://github.com/civicrm/civicrm-core/blob/f45171b377f03e8303fce6930f6bee9b35363475/CRM/Core/CodeGen/Schema.php#L3-L11